### PR TITLE
Disentangle bevy_utils/bevy_core's reexported dependencies

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Make bevy_utils less of a compilation bottleneck. Tackle #11478.

## Solution
 * Move all of the directly reexported dependencies and move them to where they're actually used. 
 * Remove the UUID utilities that have gone unused since `TypePath` took over for `TypeUuid`. 
 * There was also a extraneous bytemuck dependency on `bevy_core` that has not been used for a long time (since `encase` became the primary way to prepare GPU buffers).
 * Remove the `all_tuples` macro reexport from bevy_ecs since it's accessible from `bevy_utils`.

---

## Changelog
Removed: Many of the reexports from bevy_utils (petgraph, uuid, nonmax, smallvec, and thiserror).
Removed: bevy_core's reexports of bytemuck.

## Migration Guide
bevy_utils' reexports of petgraph, uuid, nonmax, smallvec, and thiserror have been removed. 

bevy_core' reexports of bytemuck's types has been removed. 

Add them as dependencies in your own crate instead.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Good stuff. I'd like to look into workspace-level dependencies at some point too.

File: crates/bevy_reflect/Cargo.toml [hunk 1]
```diff
@@ -10,12 +10,13 @@ keywords = ["bevy"]
 readme = "README.md"
 
 [features]
-default = []
+default = ["smallvec"]
```
Comment (NiseVoid) [inline crates/bevy_reflect/Cargo.toml:13]: Having just this feature in default feels wrong :thinking:

**Review by NthTensor — APPROVED** [review]

File: crates/bevy_reflect/Cargo.toml (hunk 1 above)
Comment (james7132) [reply crates/bevy_reflect/Cargo.toml:13]: This makes it a non-breaking change because previously it wasn't optional.

**Comment by alice-i-cecile** [comment]
CI failures are real, unsurprisingly. James, feel free to merge once you can get things building without default features.

**Pull request merged**